che-incubator/kubernetes-image-puller-operator
# Introduce webhook to prevent more than 1 KIP resource in a single namespace

Related issue: https://issues.redhat.com/browse/CRW-4980

This PR creates a validation webhook ensures that only one KIP CR can be created in a given namespace:

https://github.com/che-incubator/kubernetes-image-puller-operator/assets/83611742/09e20dad-6ec0-4a07-b326-ce2e40f26a39.

This PR also generates deployment manifests under `deploy/deployment`:
```
└── deploy/deployment/
    ├── kubernetes/
    │   ├── objects/
    │   └── combined.yaml
    └── openshift/
        ├── objects/
        └── combined.yaml
```

where manifests under `kubernetes` is meant for installation on a kubernetes cluster (requires cert-manager), and manifests under `openshift` is meant for openshift cluster (requires service-ca)

This is generated automatically when running `make deploy`.

## How to test this PR
1. Build the operator images (optional)
I've already built it here:
* `quay.io/dkwon17/kubernetes-image-puller-operator:validation`

The image can be built by running this command:
```
make docker-build IMG=quay.io/<username>/kubernetes-image-puller-operator:<tag>
```

2. In a new cluster, deploy the operator:
```
make deploy IMG=quay.io/dkwon17/kubernetes-image-puller-operator:validation
```

3. In the `default` namespace, try to create two `KubernetesImagePuller` resources in the same namespace. The webhook should prevent that from happening with an error message:

<details><summary>Screenshot</summary>

![image](https://github.com/che-incubator/kubernetes-image-puller-operator/assets/83611742/128ff770-61dd-4a58-a6cd-177259e66877)

</details>

## Testing the bundle build

With this PR, building a bundle image and deploying with `operator-sdk run bundle` works as expected:

1. Set env vars.

Note: you must have your own `kubernetes-image-puller-operator` and `kubernetes-image-puller-operator-bundle` repositories available under your quay account. 
```
$ export IMAGE_TAG_BASE=quay.io/<username>/kubernetes-image-puller-operator
$ export VERSION=1.1.0
```
3. Build and push the operator image:
```
$ make docker-build docker-push 
```
4. Build and push the bundle image:
```
$ rm -rf bundle/manifests
$ make bundle bundle-build bundle-push
```
5. Install the bundle on a cluster with OLM:
```
$ ./bin/operator-sdk run bundle $BUNDLE_IMG
```

The Kubernetes Image Puller operator should install successfully, and the webhook functionality described in the *How to test this PR* should work as well.

6. Uninstall the operator
```
./bin/operator-sdk cleanup kubernetes-imagepuller-operator
```
